pkg/grpc/universalapi/api_workflow.go

```diff
@@ -50,175 +47,196 @@ func (a *UniversalAPI) GetWorkflowAlpha1(ctx context.Context, in *runtimev1pb.Ge
+		LastUpdatedAt: timestamppb.New(response.Workflow.LastUpdatedAt),
+		RuntimeStatus: response.Workflow.RuntimeStatus,
+		Properties:    response.Workflow.Properties,
 	}
 	return res, nil
 }
 
-func (a *UniversalAPI) StartWorkflowAlpha1(ctx context.Context, in *runtimev1pb.StartWorkflowRequest) (*runtimev1pb.WorkflowReference, error) {
-	if in.WorkflowName == "" {
-		err := messages.ErrWorkflowNameMissing
-		a.Logger.Debug(err)
-		return &runtimev1pb.WorkflowReference{}, err
-	}
-
-	if in.WorkflowComponent == "" {
-		err := messages.ErrNoOrMissingWorkflowComponent
+func (a *UniversalAPI) StartWorkflowAlpha1(ctx context.Context, in *runtimev1pb.StartWorkflowRequest) (*runtimev1pb.StartWorkflowResponse, error) {
+	if err := a.validateInstanceId(in.InstanceId, true /* isCreate */); err != nil {
 		a.Logger.Debug(err)
-		return &runtimev1pb.WorkflowReference{}, err
+		return &runtimev1pb.StartWorkflowResponse{}, err
 	}
 
-	if in.InstanceId == "" {
-		err := messages.ErrMissingOrEmptyInstance
+	if in.WorkflowName == "" {
+		err := messages.ErrWorkflowNameMissing
 		a.Logger.Debug(err)
-		return &runtimev1pb.WorkflowReference{}, err
+		return &runtimev1pb.StartWorkflowResponse{}, err
 	}
 
-	workflowComponent, ok := a.CompStore.GetWorkflow(in.WorkflowComponent)
-	if !ok {
-		err := messages.ErrWorkflowComponentDoesNotExist.WithFormat(in.WorkflowComponent)
+	workflowComponent, err := a.getComponent(in.WorkflowComponent)
+	if err != nil {
 		a.Logger.Debug(err)
-		return &runtimev1pb.WorkflowReference{}, err
+		return &runtimev1pb.StartWorkflowResponse{}, err
 	}
 
-	var inputMap map[string]interface{}
-	json.Unmarshal(in.Input, &inputMap)
 	req := workflows.StartRequest{
-		InstanceID:   in.InstanceId,
-		Options:      in.Options,
-		WorkflowName: in.WorkflowName,
-		Input:        inputMap,
+		InstanceID:    in.InstanceId,
+		Options:       in.Options,
+		WorkflowName:  in.WorkflowName,
+		WorkflowInput: in.Input,
 	}
 
 	resp, err := workflowComponent.Start(ctx, &req)
 	if err != nil {
 		err := messages.ErrStartWorkflow.WithFormat(in.WorkflowName, err)
 		a.Logger.Debug(err)
-		return &runtimev1pb.WorkflowReference{}, err
+		return &runtimev1pb.StartWorkflowResponse{}, err
 	}
-	ret := &runtimev1pb.WorkflowReference{
+	ret := &runtimev1pb.StartWorkflowResponse{
 		InstanceId: resp.InstanceID,
 	}
 	return ret, nil
 }
 
 // TerminateWorkflowAlpha1 is the API handler for terminating a workflow
-func (a *UniversalAPI) TerminateWorkflowAlpha1(ctx context.Context, in *runtimev1pb.WorkflowActivityRequest) (*runtimev1pb.WorkflowActivityResponse, error) {
-	var method func(context.Context, *workflows.WorkflowReference) error
-	if in.WorkflowComponent != "" {
-		if wf, ok := a.CompStore.GetWorkflow(in.WorkflowComponent); ok {
-			method = wf.Terminate
-		}
+func (a *UniversalAPI) TerminateWorkflowAlpha1(ctx context.Context, in *runtimev1pb.TerminateWorkflowRequest) (*emptypb.Empty, error) {
+	emptyResponse := &emptypb.Empty{}
```

Comment:
Nit: this can cause an unnecessary heap allocation (and so more work for the GC) when it's not needed. I'd recommend:

```
return &emptypb.Empty{}, err
```

Reply:
There shouldn't be any unnecessary allocations because every code path returns the same `emptyResponse` value, whether there is an error or a success. Or am I misunderstanding?

Reply:
Oh I'm sorry, I missed that it was returned also in case of success.